sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -341,7 +341,7 @@ object QueryPlan extends PredicateHelper {
         // Normalize the outer references in the subquery plan.
         val normalizedPlan = s.plan.transformAllExpressions {
           case OuterReference(r) => OuterReference(QueryPlan.normalizeExpressions(r, input))
-        }
+        }.canonicalized
```

Comment:
I think this is not the right place for canonicalizing the plan in a subquery. I checked the subclasses of `PlanExpression` and it seems like that only `org.apache.spark.sql.execution.ScalarSubquery` doesn't have the canonicalized method implemented, which also makes sense given your example plans, the reuse doesn't happen when there is a scalar subquery in the plan. Could you please remove this call from here and implement the `canoncalized` method for `ScalarSubquery` and verify if it has the same effect?

Reply:
Thanks, good catch, fixed it in ScalarSubquery.